src/dataframerow/utils.jl

```diff
@@ -167,43 +185,83 @@ function row_group_slots(cols::NTuple{N,<:Union{CategoricalVector,PooledVector}}
-                refmap[missingind+1] = -1
-                refmap[missingind+2:end] .-= 1
+    refs = map(DataAPI.refarray, cols)
+    strides = (cumprod(collect(reverse(ngroupstup)))[end-1:-1:1]..., 1)::NTuple{N,Int}
+    firstinds = map(firstindex, refpools)
+    # TODO: when skipmissing=true, do not include missing values
+    # when checking whether pool is sorted
+    if sort && !all(issorted, refpools)
+        # Compute vector mapping missing to -1 if skipmissing=true
+        refmaps = map(cols, refpools) do col, refpool
+            refmap = collect(0:length(refpool)-1)
+            if skipmissing
+                fi = firstindex(refpool)
+                missingind = findfirst(ismissing, refpool)
+                if missingind !== nothing
+                    mi = something(missingind)
+                    refmap[mi-fi+1] = -1
+                    refmap[mi-fi+2:end] .-= 1
+                end
+                if sort
+                    nm = missingind === nothing ? eachindex(refpool) :
+                        setdiff(eachindex(refpool), missingind)
+                    perm = sortperm(view(refpool, nm))
+                    invpermute!(view(refmap, nm .- fi .+ 1), perm)
+                end
+            elseif sort
+                # FIXME: collect is needed for CategoricalRefPool
+                perm = sortperm(collect(refpool))
+                invpermute!(refmap, perm)
             end
+            refmap
         end
-        refmap
-    end
-    strides = (cumprod(collect(reverse(ngroupstup)))[end-1:-1:1]..., 1)::NTuple{N,Int}
-    @inbounds for i in eachindex(groups)
-        local refs
-        let i=i # Workaround for julia#15276
-            refs = map(c -> c.refs[i], cols)
+        @inbounds for i in eachindex(groups)
+            local refs_i
+            let i=i # Workaround for julia#15276
+                refs_i = map(c -> c[i], refs)
+            end
+            vals = map((m, r, s, fi) -> m[r-fi+1] * s, refmaps, refs_i, strides, firstinds)
+            j = sum(vals) + 1
+            # x < 0 happens with -1 in refmap, which corresponds to missing
+            if skipmissing && any(x -> x < 0, vals)
+                j = 0
+            else
+                seen[j] = true
+            end
+            groups[i] = j
         end
-        vals = map((m, r, s) -> m[r+1] * s, refmaps, refs, strides)
-        j = sum(vals) + 1
-        # x < 0 happens with -1 in refmap, which corresponds to missing
-        if skipmissing && any(x -> x < 0, vals)
-            j = 0
-        else
-            seen[j] = true
+    else
+        missinginds = map(refpools) do refpool
+            something(findfirst(ismissing, refpool), lastindex(refpool)+1)
+        end
+        @inbounds for i in eachindex(groups)
+            local refs_i
+            let i=i # Workaround for julia#15276
+                refs_i = map(refs, missinginds) do ref, missingind
+                    r = Int(ref[i])
+                    if skipmissing
```

Comment:
can we have `skipmissing` branch outside? In this way for the most common case of `skipmissing=false` we have an unbranching code here (or you leave it as is as you assume that the compiler should optimize this out?).

Reply:
In my tests this doesn't make any difference to performance. In general I think LLVM is pretty good at hoisting this kind of check.

Reply:
Great - thank you!